### DISCUSSION TITLE
PCB progress

### DISCUSSION THREAD

**Discussion by jimbro1000** [reifsnyderb/Apple-II-R0#21] [Show and tell]
![AppleII](https://github.com/reifsnyderb/Apple-II-R0/assets/36511452/3e3016fb-38dc-4976-98cb-960bc1edd719)
PCB design is nearly done - component positions is firmed up, routing is completed but the labelling needs refinement as the left-overs from the original gerbers are making a mess

**Comment by reifsnyderb** [comment]
The PCB is looking great!  Did you start out using Mike Willegal's gerber files?  I have them but had problems importing them into KiCAD with the proper trace widths.  Electrically, my import is ok.  But the width's are all too narrow.

**Reply by jimbro1000** [reply]
I started from the same gerbers but something in them is messed up with the clearance margin showing as additional width instead of a void. I just mapped the two copper layers onto unused layers in the editor and the silkscreen onto another because 50% of it had the same issue.

It's worked out better in the end as it has forced me to verify all the traces against the schematic and bind it all appropriately. This is all native KiCad now too so much easier to work with. Only needs 3 more rows and the fiddly video circuit to finish

**Comment by jimbro1000** [comment]
Once this is done there will be a rather tedious job of merging the fork, a simple pull request is unlikely to work and KiCad doesn't do merges particularly nicely. We can deal with that when it comes up though.

**Reply by reifsnyderb** [reply]
It may not be too bad.  When the PCB is done, let me know and I'll try to swap the PCB file in and see what happens.

**Comment by reifsnyderb** [comment]
I've got some R0 replica boards made and they show the same issue on H14 in regards to pin 6.  I'd love to have an original R0 handy to compare.

**Reply by jimbro1000** [reply]
I saw you had a couple left on eBay (not stalking, just trying to find another way to contact you). 

I've no feedback on how well those boards work. I fancied building one but the kits were just nastily priced so here I am, trying to make it cheaper

**Reply by reifsnyderb** [reply]
Those boards were made using Mike Willegal's gerbers as he has posted on his website.  I got them made figuring on building one but decided that having to round up all of the original parts could be quite expensive.  Since the minimum order is 5, it made sense to ebay the rest.

So far, it looks like H14 is the only issue.  I just sent Mike a message to see if he knows about H14.  All other signs point to those board working fine.

Once we get this all sorted out, I figure the next step is to figure out how to convert the Apple II over to all commonly available components.  I think I have this all worked out in the schematic.  It would also make sense to add all of the upgrades provided by the Apple Plus as well.

**Reply by jimbro1000** [reply]
That was my goal too - top of the list is switching the RAM over to SRAM and chopping out the whole RAS/CAS address multiplexing. Just needs something to isolate data-in and data-out. That should pretty much free up two rows on the board (at least)

Everything else is regular LS logic although some of those may be getting hard to find even mixing HCT in with it.

There are two logic chips I've not looked into that aren't regular 74 series

Everything else I've checked as available from Digikey

**Reply by jimbro1000** [reply]
Just thought I'd take a look into those two chips - the first is the 8T97 and the second entry in the search results is an old forum post from Mike saying use a 74LS367 instead...

Was expecting much the same for the 8T28 but it turns out there isn't a direct swap. A 74LS240/241 can perform the same task but isn't pin compatible (20 pins instead of 16). Given the physical location on the board this shouldn't be too drastic to fix

**Reply by reifsnyderb** [reply]
There are so many changes required, I am not concerned about pin compatibility.  Only whether it works or not and that the chips are available.

I was thinking about replacing the HT97's and HT28's with 74AHCT245's as they are available and cheaper.  

For the SRAM, 48k of a AS6C1008 could be used.  The ROM could be replaced by an SST39SF010.  While there is a lot of wasted space, those chips are readily available and cheap.  The character ROM should be able to also be replaced by an SST39SF010.

It might be possible to even add some additional logic to use another 16k, of the SRAM, for upper memory memory.  Just something to consider.

I think I've got the multiplexer worked out for the SRAM, as well.  A single 74F541 would be needed to separate the AS5C1008 from the data bus for writes.

558's are no longer available but 2 556's can be used instead.  

PLD's, such as the ATF16V8 and ATF22V10 can be used as well.  I think I've got pseudo code worked out for the PLD's as well.

The total chip count reduction should be around 50%.

I need to update the schematic with any Rev 0 changes you found and I'll post it.  Right now, I've got the board partially laid out but am not happy with it.  I think it would be a good idea to just accept that there are a lot of changes and re-do most of the board.  Most likely, the best thing to do is to keep the back of the board, by the slots, similar and re-do the rest as there are so many changes to deal with.

**Reply by jimbro1000** [reply]
I'd refer you to the work I've done on the Dragon 32. Step 1 was reproduce it with minimal changes. Step 2 redesign it without radically changing the schematics. Step 3 throw it all away and start again

I have 3 different repositories for each step as they are 3 very different boards. The first one just works as a direct replacement. The second I separated the video circuits onto a stacked board so I could have a single main board design instead of different ones for each video format - that board works but the PAL video board is giving me a headache...

The third board throws away all of the legacy parts and opens the design up for heavy rework - the SAM chip is now a CPLD and I'm in the middle of redesigning the VDG, first real tests of that should start this week (so I want to get this Apple II board finished today). I've reworked the RAM circuits to use SRAM and the CPLD SAM means no more multiplexing needed but it still does all the memory mapping and timing

**Reply by reifsnyderb** [reply]
Well, I've been through three attempts and two schematics on the new board.  I'd have to count them, but I think there's over 15 revisions in total.  I'll soon review the latest revision and compare any issues to what you have found so as to update the board.

With the exception of adding more support for PAL, I think I've got the schematic pretty close.  The functionality is that of an Apple II Plus so as to have the reset feature, extra color support, and other minor changes.

What are you using for a CPLD?  I like CPLDs, but software availability, and software toolchain complexities, and the desire to not deal with level shifters, have resulted in my deciding to use PLD chips.

**Reply by jimbro1000** [reply]
The CPLD in this case is a Xilinx XC95144XL, operates at 3.3V but its is 5v tolerant on IO. The conversion project isn't mine - I've just forked it, stripped out the stuff I don't need and added in a few more extras. The original project is here: https://gitlab.com/sixxie/samx4

**Reply by reifsnyderb** [reply]
Isn't that TRS-80 related?  Or is it close enough to use in an Apple II.

**Reply by jimbro1000** [reply]
Nothing to do with a TRS-80 as in the Z80 based machines - it is used in the Color Computer, and in the Dragon computers which is where I come in and keep finding ways to make it better

In terms of using it in an Apple - that wasn't what I was thinking at all. That said the address mapping logic, RAM address multiplexing and timing could all be wrapped up in the same way - it would remove about 50% of the chips on the board (assuming you convert the 4116 RAM to SRAM and the ROMs to a single eprom or flash rom). It just leaves the video circuit but by the time you've gone that far you might as well encode the video circuit too given it is just a handful of logic gates and then you're left with a CPU, a ROM, character generator (ROM), a couple of SRAM (or maybe just one), a CPLD and the discrete components. At that point you could probably chop the lower half of the board off given the amount of free space it would create. That would still make for an interesting project but really doesn't lee much of the original concept behind

**Reply by reifsnyderb** [reply]
My schematic does just that.  I replaced all of the individual logic gates with PLDs.  Even the video has been replaced with PLDs.  Even so, the PLDs are easy to program.

I really started working on this because I saw the Commander X16 and the new Z80 projects.  While they are neat, there is little software available and they really aren't too hackable due to the use of CPDs and FPGAs.  Since the Apple II had a lot of software available and didn't use any LSI chips, it seemed to be a good idea to just "modernize" the Apple II.  I figured this would be tied in with verifying the existing schematics with a known working system board design.  The end result will be a well understood design that is cheaper, 100% compatible with existing software, able to be understood by one person, and can be built with chips that are all rather cheap.

Using the PLDs will reduce the chip count, board space, simplify the board traces, and still be easily modified.  The same burner used for programming the PLDs can also program the ROMs.  There is no concern about dealing with JTAG interfacing and the software required to program CPLDs.  It is a compromise over just using logic gates, but I think it is pretty reasonable.  If PLDs, for some reason, are no longer available in the future, it would be easy to convert back to glue logic.

I think an end goal of having an Apple II Plus compatible computer built all with cheap and current components would be a good compromise.

**Reply by jimbro1000** [reply]
All sounds very reasonable, and I can definitely see it's a good way forwards.

I will have to look forward to seeing your solution, I really need to pick a tool chain and stick with it, right now I'd be a bit lost converting it all reliably. Still learning as I go

**Reply by reifsnyderb** [reply]
Here's what I came up with for the RAM Address multiplexing....

[AppleII-RAM Address Multiplexing.pdf](https://github.com/reifsnyderb/Apple-II-R0/files/14159352/AppleII-RAM.Address.Multiplexing.pdf)

I'll have to double-check everything, of course.  But I am pretty certain it will work.

**Reply by jimbro1000** [reply]
Ah cool! I was thinking about the need for multiplexers in the design - whether it is better to use discrete chips or try to include them in the PLD. If you're using small PLDs then it is a no brainer as the multiplexers conveniently reduce the amount of IO pins required. Seems my thinking wasn't too far off

**Comment by jimbro1000** [comment]
![AppleII](https://github.com/reifsnyderb/Apple-II-R0/assets/36511452/b1c72dcb-b9b6-4f69-bbce-a0b48ff7d886)
I think that is done now... I haven't identified the SOFT5 pins that need separating but it is purely cosmetic

Mounting holes added, along with the prototyping area in the bottom right

**Reply by reifsnyderb** [reply]
The board is looking great!

**Comment by jimbro1000** [comment]
I've added the two test points to the board and there are (according to DRC) zero errors! (Ignoring all the overlapping silk screen and margin conflicts)
Just bringing the schematics into alignment with yours so you should be able to just add the PCB files to your schematics without breaking anything

**Comment by jimbro1000** [comment]
I've made a little progress over the last week - trying to merge the schematics to a point where the PCB can be lifted and shifted between them.
Also thrown together an amended version that provides two footprints for the char-gen rom so a 2716 or 2816 can be used instead without a shim board - this isn't on the main schematic yet

**Reply by reifsnyderb** [reply]
Sounds great!  I'll take a look a little later.  I think the amended version, with the 2716 or 2816, should be a separate project.  That way, we'll have the original, and the "improved" version.

**Reply by jimbro1000** [reply]
The "enhanced" version is on a separate branch for that very reason

**Reply by reifsnyderb** [reply]
I took a look at the files.  You made lots of progress!  I added the Apple Computer Inc. text and changed a small number of minor footprints so they look more authentic.  (i.e.  User-1, Test points, the 16 pin keyboard connectors, etc.)

**Reply by reifsnyderb** [reply]
Here's a picture of the updated PCB:

![AppleII-populated](https://github.com/reifsnyderb/Apple-II-R0/assets/46803603/19d40929-3ad8-4a61-b5f2-8e104a961a0a)

**Reply by jimbro1000** [reply]
Nice - I'll stop tinkering on the "original" board then
